File: tut04/client2.py
import socket

def main():
    try:
        # Create a socket object
        client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        # Server IP address and port
        host = "127.0.0.1"  # Server IP address (localhost)
        port = 12345       # Server port

        # Connect to the server
        client_socket.connect((host, port))

        # Send a string to the server
        message = input("Enter a string: ")
        client_socket.send(message.encode())

        # Receive the count of vowels from the server
        vowel_count = client_socket.recv(1024).decode()
        print(f"Server response: The number of vowels in the string is {vowel_count}")

    except ConnectionRefusedError:
        print("Connection to the server was refused.")
    except Exception as e:
        print(f"An error occurred: {str(e)}")
    finally:
        # Close the connection
        client_socket.close()

if __name__ == "__main__":
    main()

File: tut05/client2.cpp
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>

const int BUFFER_SIZE = 1024;

int main() {
    // Create a socket
    int clientSocket = socket(AF_INET, SOCK_STREAM, 0);
    if (clientSocket == -1) {
        perror("Error creating socket");
        exit(EXIT_FAILURE);
    }

    // Server address and port
    struct sockaddr_in serverAddr;
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(8080);                   // Server port
    serverAddr.sin_addr.s_addr = inet_addr("127.0.0.1"); // Server IP address

    // Connect to the server
    if (connect(clientSocket, reinterpret_cast<struct sockaddr *>(&serverAddr), sizeof(serverAddr)) == -1) {
        perror("Error connecting to server");
        exit(EXIT_FAILURE);
    }

    // Send a message to the server
    char message[BUFFER_SIZE];
    printf("Enter the message to be sent to the server: ");
    scanf("%s", message);
    int messageLength = strlen(message);

    // Send the message
    int bytesSent = send(clientSocket, message, messageLength, 0);
    if (bytesSent == -1) {
        perror("Error sending data");
        close(clientSocket);
        exit(EXIT_FAILURE);
    }

    // Receive a response from the server
    char buffer[BUFFER_SIZE];
    int bytesRead = recv(clientSocket, buffer, BUFFER_SIZE - 1, 0);
    if (bytesRead == -1) {
        perror("Error receiving data");
        close(clientSocket);
        exit(EXIT_FAILURE);
    }

    // Null-terminate the received data and print
    buffer[bytesRead] = '\0';
    printf("Received response from server: %s\n", buffer);

    // Close the socket
    close(clientSocket);

    return 0;
}
